### PR TITLE
fix: [DEL-2855]: Add API to fetch delegates with primary version.

### PR DESCRIPTION
**Pull request by gaurav-jain1**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)

### CODE REVIEW

File: 400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java [hunk 1]
```diff
@@ -212,6 +212,23 @@ public DelegateSetupResourceV3(DelegateService delegateService, DelegateScopeSer
...
+  @Timed
+  @ExceptionMetered
+  @PublicApi
+  @Operation(operationId = "getActiveDelegatesWithPrimary",
+      summary = "List all AccountId along with corresponding DelegateIDs in primary version.",
+      responses =
+      {
+        @io.swagger.v3.oas.annotations.responses.ApiResponse(responseCode = "default",
+            description = "A list of accountID along with their DelegateIDs having primary version.")
+      })
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java:226]: we should not expose account ids in a rest api

Comment (gaurav-jain1) [reply 400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java:226]: What's the consequences ? These are just IDs being sent for further investigation. 
This API is under auth anyway.

**Review by mdmilic — COMMENTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java, line 597 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuPZr3P2Z3OFTCHDVmB:-MuPZr3P2Z3OFTCHDVmC:be99ge2) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L597)):*
> ```Java
>   public Map<String, List<String>> getActiveDelegatesWithPrimary() {
>     String primaryVersion = Arrays.stream(configurationController.getPrimaryVersion().split("-")).findFirst().get();
>     Map<String, List<String>> delegatesPerVersion =
> ```

These are delegates per account. Rename
___
*[400-rest/src/main/java/software/wings/service/intfc/DelegateService.java, line 74 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuPZUzP1OUY14nt0zi_:-MuPZUzP1OUY14nt0zia:b4hbb62) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/service/intfc/DelegateService.java#L74)):*
> ```Java
>   Double getConnectedRatioWithPrimary(String targetVersion);
> 
>   Map<String, List<String>> getActiveDelegatesWithPrimary();
> ```

All the naming 'withPrimary' should be changed to something better (including API). It's not clear from the names what primary means. 
___
*[420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java, line 68 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuPZCJH6B415Yxgidnx:-MuPZCJH6B415Yxgidny:b-wrnfki) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java#L68)):*
> ```Java
>   }
> 
>   public Map<String, List<String>> obtainActiveDelegatesPerVersion(String version) {
> ```

My understanding is that this returns delegates per account for some specific version. Should be named accordingly.

**Review by mdmilic — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuP-2PqYP0e3kvDjkZy-r1-226:-MuP_j4RB36DrMFLRh8k:b6cj0jn) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

What's the consequences ? These are just IDs being sent for further investigation. 
This API is under auth anyway.
</blockquote></details>

This seems like an admin endpoint, so should go in the admin resource and also have appropriate auth. We should definitely not make this a public API (leaking customers information to other customers).

**Comment by raghuAtWings** [comment]
___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuP-2PqYP0e3kvDjkZy-r1-226:-MuRWSR484m7pOTIZQ_i:b7d07g0) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

This seems like an admin endpoint, so should go in the admin resource and also have appropriate auth. We should definitely not make this a public API (leaking customers information to other customers). 
</blockquote></details>

also i see two endpoints, one in this file and one in 400-rest/src/main/java/software/wings/resources/DelegateSetupResource.java

why do we need two endpoints?

**Comment by raghuAtWings** [comment]
___
*[420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java, line 68 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuPZCJH6B415Yxgidnx:-MuRWfAu25zrLCqGKYC0:b32v57) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/420-delegate-service/src/main/java/software/wings/service/impl/DelegateConnectionDao.java#L68)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

My understanding is that this returns delegates per account for some specific version. Should be named accordingly.
</blockquote></details>

+1

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 6 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuP-2PqYP0e3kvDjkZy-r1-226:-MuVILTU6L5qMb91qmFz:b2j5qcf) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

also i see two endpoints, one in this file and one in 400-rest/src/main/java/software/wings/resources/DelegateSetupResource.java

why do we need two endpoints?
</blockquote></details>

Done. Added a check for UserID.

Raghu- Not sure about this, I just followed previous API call 'connected-ratio-with-primary' , since that's defined at both places too,



_Code quote:_
```Java
delegates-with-primary
```

**Comment by gaurav-jain1** [comment]
trigger messagecheck

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions.

**Comment by raghuAtWings** [comment]
___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuP-2PqYP0e3kvDjkZy-r1-226:-Mume8q7E826FrdRzrsA:b-lfz6nh) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Done. Added a check for UserID.

Raghu- Not sure about this, I just followed previous API call 'connected-ratio-with-primary' , since that's defined at both places too,
</blockquote></details>

we should just have one end point and work with ops folks to use the right api

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: Some of the checks are too old. 1 change requests. 1. unresolved discussions.

**Review by raghuAtWings — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: Some of the checks are too old. 2 change requests. 1. unresolved discussions. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuP-2PqYP0e3kvDjkZy-r1-226:-MumeJOM0fPl25X8CfRI:b-acdoyr) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

we should just have one end point and work with ops folks to use the right api
</blockquote></details>

lets get rid of one

**Review by abhinav-harness — DISMISSED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 2 change requests. 1. unresolved discussions.

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 2 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r1](https://reviewable.io/reviews/harness/harness-core/30280#-MuP-2PqYP0e3kvDjkZy-r1-226:-MusutRmAVb71CvMZN63:b-krjf53) ([raw file](https://github.com/harness/harness-core/blob/28969eb38f6589d0bfbdfd2bca8ba2fc4d879d91/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

lets get rid of one
</blockquote></details>

Done. I'll keep v3 one.

**Review by abhinav-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280#-:-MutWbmv55vGMfDN6erV:bnfp4nl)* status: 2 change requests. 1. unresolved discussions.

**Review by gaurav-jain1 — COMMENTED** [review]
> Reviewed 4 of 6 files at r2, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 2 change requests. 1. unresolved discussions.

**Review by gaurav-jain1 — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280#-:-MutY9ZU6SgNnKPuPd7_:bg7igk9)* status: 2 change requests. 1. unresolved discussions.

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions.

**Review by mdmilic — APPROVED** [review]
:lgtm:
There might be still a place or two (possibly even in the pre existing code) where the name of the API or method is e.g. `delegateWithPrimary()`. For me this is not really understandable (primary what?), i.e. the name doesn't have the object of the sentence (`primary` being adjective with no object), but looks good overall. 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280#-:-Muwy8Er6YfmejEekjrS:bwvbri4)* status: :shipit: complete! ✓

**Review by mdmilic — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: :shipit: complete! ✓

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r3](https://reviewable.io/reviews/harness/harness-core/30280#-Mux-1QD3eHISDlCz0_6:-Mux-1QD3eHISDlCz0_7:b-9cwf0i) ([raw file](https://github.com/harness/harness-core/blob/ae769890e14cd81519d58a66b8438808f2563f14/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
> ```Java
>   @Timed
>   @ExceptionMetered
>   @PublicApi
> ```

Hey I still don't think this should be a public API

**Review by mdmilic — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: :shipit: complete! ✓

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r3](https://reviewable.io/reviews/harness/harness-core/30280#-Mux-1QD3eHISDlCz0_6:-Mux-VujCOJvy6ZPcLm6:b-afpxb) ([raw file](https://github.com/harness/harness-core/blob/ae769890e14cd81519d58a66b8438808f2563f14/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Hey I still don't think this should be a public API
</blockquote></details>

And also should probably go to some Admin...Resource.java

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions. 

___
*[400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java, line 226 at r3](https://reviewable.io/reviews/harness/harness-core/30280#-Mux-1QD3eHISDlCz0_6:-MuyW5r99xtfLzmjQvU_:br820q2) ([raw file](https://github.com/harness/harness-core/blob/ae769890e14cd81519d58a66b8438808f2563f14/400-rest/src/main/java/software/wings/resources/DelegateSetupResourceV3.java#L226)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

And also should probably go to some Admin...Resource.java 
</blockquote></details>

@mdmilic Can you be specific how to make the API private, providing some examples will be better. 
As per my understanding I have put `harnessUserGroupService` check to avoid calling by unauthorized users.



_Code quote:_
```Java
harnessUserGroupService.isHarnessSupportUser(UserThreadLocal.get().getUuid()))
```

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions.

**Review by gaurav-jain1 — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions.

**Review by gaurav-jain1 — COMMENTED** [review]
> Reviewed 3 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 3 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30280)* status: 1 change requests. 1. unresolved discussions.

**Pull request merged**